From 53d64899632db6165e073a5a43720557d6710661 Mon Sep 17 00:00:00 2001
From: Rainer Gerhards <rgerhards@adiscon.com>
Date: Tue, 10 Jun 2025 18:25:30 +0200
Subject: [PATCH] omusrmsg: use utmpx on systems where available

- this is driven by Solaris: it does not provide UT_LINESIZE
- prefer utmpx when available and map setutent()/getutent()
- derive UT_LINESIZE from platform macros to compile on Solaris
- if nothing else helps, set UT_LINESIZE to 32
  Note that 32 is a somewhat save default, but not perfect.

Fixes https://github.com/rsyslog/rsyslog/issues/5678
---
 tools/omusrmsg.c | 43 +++++++++++++++++++++++++++++++------------
 1 file changed, 31 insertions(+), 12 deletions(-)

diff --git a/tools/omusrmsg.c b/tools/omusrmsg.c
index 0066c7e9cd..09e0f5c335 100644
--- a/tools/omusrmsg.c
+++ b/tools/omusrmsg.c
@@ -8,7 +8,7 @@
  * File begun on 2007-07-20 by RGerhards (extracted from syslogd.c, which at the
  * time of the fork from sysklogd was under BSD license)
  *
- * Copyright 2007-2018 Adiscon GmbH.
+ * Copyright 2007-2025 Adiscon GmbH.
  *
  * This file is part of rsyslog.
  *
@@ -35,14 +35,27 @@
 #include <signal.h>
 #include <ctype.h>
 #include <sys/param.h>
-#ifdef HAVE_UTMP_H
-#  include <utmp.h>
-#  define STRUCTUTMP struct utmp
-#  define UTNAME ut_name
-#else
-#  include <utmpx.h>
-#  define STRUCTUTMP struct utmpx
-#  define UTNAME ut_user
+#ifdef HAVE_UTMPX_H
+# include <utmpx.h>
+# define STRUCTUTMP struct utmpx
+# define UTNAME ut_user
+# define setutent setutxent
+# define getutent getutxent
+# define endutent endutxent
+# ifndef UT_LINESIZE
+/* __UT_LINESIZE for glibc; _UTX_LINESIZE common on Solaris */
+#  ifdef __UT_LINESIZE
+#   define UT_LINESIZE __UT_LINESIZE
+#  elif defined(_UTX_LINESIZE)
+#   define UT_LINESIZE _UTX_LINESIZE
+#  else /* method of last resort */
+#   define UT_LINESIZE 32
+#  endif
+# endif
+#elif defined(HAVE_UTMP_H)
+# include <utmp.h>
+# define STRUCTUTMP struct utmp
+# define UTNAME ut_name
 #endif
 #include <unistd.h>
 #include <sys/uio.h>
@@ -173,9 +186,15 @@ ENDdbgPrintInstInfo
 #ifdef OS_BSD
 /* Since version 900007, FreeBSD has a POSIX compliant <utmpx.h> */
 #if defined(__FreeBSD__) && (__FreeBSD_version >= 900007)
-#  define setutent(void) setutxent(void)
-#  define getutent(void) getutxent(void)
-#  define endutent(void) endutxent(void)
+#  ifndef setutent
+#    define setutent() setutxent()
+#  endif
+#  ifndef getutent
+#    define getutent() getutxent()
+#  endif
+#  ifndef endutent
+#    define endutent() endutxent()
+#  endif
 #else
 static FILE *BSD_uf = NULL;
 void setutent(void)
